### PR TITLE
Caching homepage to decrease TTFB

### LINKED ISSUE #1553: Cache entire homepage to reduce high TTFB
The home page currently has a time to first byte (TTFB) [of almost 6 seconds](https://www.webpagetest.org/result/181107_6D_b5d5d65a5c06ce14ee0eefa12a5a4332/).
Every time a user hits this page, many Python functions must execute in different parts, hitting solr numerous times.

Each of these solr hits is cached, but in the case of this page, it might make much more sense to cache the final HTML.

Since memcached is the only caching layer we have right now we'll store the generated HTML for the homepage in memcached.

To be extremely conservative, the cache time should be 5 minutes - as if the loan status of a book changes, a user will think the book's still available.

After this change we will investigate the impacts on memcached memory and TTFB. From this information, we'll have learned about how caching helps our TTFB and improves user performance.

### PR DESCRIPTION
**Pull request by mekarpeles**
## Description

Closes #1553 

## Testing

We'll want to test this on local and dev and get @hornc's approval that it's behaving correctly on dev (seeing occasional traceback messages which I'd like another pair of eyes on)

## Evidence

I was seeing ~300ms load-time

### CODE REVIEW

**Comment by mekarpeles** [comment]
@jdlrobson the page loads quickly, js seems to be behaving funky -- `vendor-v2.js` taking a very long time to load on dev (13.2 sec)

**Comment by hornc** [comment]
Testing in local dev environment gives errors on loading the main page:

```
web_1        | Traceback (most recent call last):
web_1        |   File "/usr/local/lib/python2.7/dist-packages/web/application.py", line 239, in process
web_1        |     return self.handle()
web_1        |   File "/usr/local/lib/python2.7/dist-packages/web/application.py", line 230, in handle
web_1        |     return self._delegate(fn, self.fvars, args)
web_1        |   File "/usr/local/lib/python2.7/dist-packages/web/application.py", line 462, in _delegate
web_1        |     return handle_class(cls)
web_1        |   File "/usr/local/lib/python2.7/dist-packages/web/application.py", line 438, in handle_class
web_1        |     return tocall(*args)
web_1        |   File "/openlibrary/infogami/utils/app.py", line 189, in <lambda>
web_1        |     HEAD = GET = POST = PUT = DELETE = lambda self: delegate()
web_1        |   File "/openlibrary/infogami/utils/app.py", line 208, in delegate
web_1        |     return getattr(cls(), method)(*args)
web_1        |   File "/openlibrary/openlibrary/plugins/openlibrary/home.py", line 62, in GET
web_1        |     return web.template.TemplateResult(get_cached_homepage())
web_1        |   File "/openlibrary/openlibrary/plugins/openlibrary/home.py", line 53, in get_cached_homepage
web_1        |     get_homepage, "home.homepage", timeout=5*dateutil.MINUTE_SECS)()
web_1        |   File "/openlibrary/openlibrary/core/cache.py", line 106, in __call__
web_1        |     value, t = self.update(*args, **kw)
web_1        |   File "/openlibrary/openlibrary/core/cache.py", line 148, in update
web_1        |     self.memcache_set(args, kw, value, t)
web_1        |   File "/openlibrary/openlibrary/core/cache.py", line 190, in memcache_set
web_1        |     json = self.json_encode([value, time])
web_1        |   File "/openlibrary/openlibrary/core/cache.py", line 181, in json_encode
web_1        |     return simplejson.dumps(value, separators=(",", ":"))
web_1        |   File "/usr/local/lib/python2.7/dist-packages/simplejson/__init__.py", line 399, in dumps
web_1        |     **kw).encode(obj)
web_1        |   File "/usr/local/lib/python2.7/dist-packages/simplejson/encoder.py", line 296, in encode
web_1        |     chunks = self.iterencode(o, _one_shot=True)
web_1        |   File "/usr/local/lib/python2.7/dist-packages/simplejson/encoder.py", line 378, in iterencode
web_1        |     return _iterencode(o, 0)
web_1        |   File "/usr/local/lib/python2.7/dist-packages/simplejson/encoder.py", line 273, in default
web_1        |     o.__class__.__name__)
web_1        | TypeError: Object of type TemplateResult is not JSON serializable
```
I'm looking further to see what else I can gather.

**Comment by bfalling** [comment]
LGTM

File: openlibrary/plugins/openlibrary/home.py [hunk 1]
```diff
@@ -32,25 +32,34 @@
...
+    except Exception:
+        logger.error("Error in getting stats", exc_info=True)
+        stats = None
+    blog_posts = get_blog_feeds()
+    page = render_template(
+        "home/index", stats=stats,
+        blog_posts=blog_posts,
+    )
+    page.v2 = True
+    return page
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/home.py]: @hornc perhaps the cache is having trouble serializing page to json and instead we should be returning page.dict() or dict(page)?

**Comment by hornc** [comment]
LGTM -- code review and tested in local docker env :+1: no errors, and css results appear cached as expected

File: openlibrary/plugins/openlibrary/home.py [hunk 2]
```diff
@@ -32,25 +32,44 @@
...
-        page = render_template(
-            "home/index", stats=stats,
-            blog_posts=blog_posts,
-        )
-        page.v2 = True
-        return page
+        cached_homepage = get_cached_homepage()
+        # when homepage is cached, home/index.html template
+        # doesn't run ctx.setdefault to set the bodyid so we must do so here:
+        web.template.Template.globals['ctx']['bodyid'] = 'home'
```
Comment (mekarpeles) [inline openlibrary/plugins/openlibrary/home.py]: This line should be replaced with `delegate.context.setdefault('bodyid', 'home')`

**Pull request merged**